### ISSUE TITLE
Provide a means of adding multiple simulations through the API

### ISSUE THREAD

**Issue by JohnFDavenport** [SpectoLabs/hoverfly#552]
The use-case is I want to be able to re-use existing functional tests for later-stage testing, especially performance testing. That means I will want to load multiple simulations into Hoverfly and then create a test that uses them all.

I do not want to have to compose a valid simulation file myself offline. That would force me to take a number of valid simulation files and compose them. That seems a lot of work at best and at worst in all probability, the simulation would be invalid.

The existing [API PUT Simulation](https://hoverfly.readthedocs.io/en/latest/pages/reference/api/api.html#put-api-v2-simulation) will always overwrite any existing simulation and is therefore not suitable.

**Comment by benjih** [comment]
The way I think I would implement this is to extend the simulation API to accept POST requests which would add the new simulation to the current simulation.

What do you think of this @mogronalol ?

**Comment by tommysitu** [comment]
For tools like Mockito, the default mode is always append, and it will not delete your mocking until reset() is called. If I mock the same request again, it will override the existing one. Maybe we can adopt this pattern to hoverfly, so:
1. Post simulation always append 
2. Capturing also append (we actually hack this multi-capture function in hoverfly-java)
3. A new request will override the old one if they are identical.

**Comment by mogronalol** [comment]
I'd agree with @tommysitu, but I wonder if we need more API endpoints. Something like

`PUT /api/v1/simulation` sets whole simulation
`POST /api/v1/simulation` merges simulations together - this might not be as simple as we like
`POST /api/v1/simulation/data/pairs` appends pairs

**Comment by JohnFDavenport** [comment]
Yes, 100% with @tommysitu  although he may mean Wiremock not Mockito? We, therefore, all agree that POST simulation always appends. A PUT would reset.

A bank of green lights on this much.

Two questions on the need for anything else for @mogronalol :

1. I think every append is a merge in the sense that Hoverfly will always prevent duplicate requests (i.e. they should be discarded with a warning). We do not allow duplicates in Hoverfly at this time. (Caveate: when we add support for stateful stubs we might do something, maybe like Mirage, but that's another subject entirely).
2. A well-formed simulation file consists of N pairs, so isn't a simulation with N=1 pairs good enough?

**Comment by JohnFDavenport** [comment]
Your thoughts @danielbryantuk ?

**Comment by danielbryantuk** [comment]
My 2 cents (after chatting with @mogronalol and @JohnFDavenport):

Because the 'simulation' resource always implicitly exists, a PUT replacing the whole simulation is good.

I think POSTing to the (already existing) simulation resource would technically (in HTTP terms) not be valid, and as @mogronalol has pointed out above - it would also be difficult to do a merge

I like the idea of POSTing pair data to 'simulation/data/pairs' too (on a side note, do the pairs get an ID? If so, they then could be deleted with DELETED 'simulation/data/pairs/<id>'?

So, in general, a 👍  to the ideas already suggested :-)

**Comment by tommysitu** [comment]
I have one use case for this. When I am testing a component which consumes a variety of external APIs. I can build simulations with DSL for some (such as internal microservices) but need to do capturing for others (such as GitHub API). In which case, I need a `HoverflyRule` in JUnit test that can composite a simulation from multiple sources. For example: 

```
hoverflyRule.addSimulation(dsl(...));
hoverflyRule.addSimulation(defaultPath(...));   // from previous captured data
```

**Comment by konjp** [comment]
Any update on this feature? Use case with `HoverflyRule`  could be very useful.

**Comment by konjp** [comment]
@tommysitu thanks on fast reply. Similar to use case you described, when testing component with multiple external sources, it will be nice that e.g. simulation for every single external service can be created and combine later when needed. That will also add more to better project organization.

**Comment by disfluxly** [comment]
My temporary answer to this has been to write a simple python script that combines multiple simulation files into a single simulation file. Essentially I'm just combining the "pairs" into a single array.

Currently I'm storing all simulations in a single GitLab Project. Then each application has a simulations.txt that contains a list of needed simulations as relative paths to the GitLab Project. Once I've created a single simulation file for all of an application's web dependencies, I push it into a Hoverfly Docker Image and push it to the GitLab registry where the application source code is, which makes it a bit simpler for developers to use.

**Comment by benjih** [comment]
Having let some time pass to think about this. I think putting something like into Hoverfly is a bad idea. I don't think it should be Hoverfly's responsibility to manage merging simulations. I anticipate that on some cases, merging simulations may require user interaction to resolve conflicts. Supporting something like this would vastly complicate the simulation interface that current exists for importing.

I like @disfluxly's approach by managing the move outside of Hoverfly and then using the resulting simulation with Hoverfly. I do feel like we should provide something that does this rather than expecting users to write scripts to do it for them.

I'm imagining a simulation merge tool, unsure if its something that should be on `hoverctl` or should be another tool.

**Comment by JohnFDavenport** [comment]
To clarify, we think there is a need for number of converters and other utilities. Including them in Hoverfly itself would bloat the codebase and undermine it as a fast API simulation/Service Virtualization tool. If there are any third-party utilities out there that are OSS let us know.

**Comment by JohnFDavenport** [comment]
Well, what a palava! It looks like this is possible on import so it ought to be documented!

**Comment by tommysitu** [comment]
Fixed in [v1.0.1](https://github.com/SpectoLabs/hoverfly/releases/tag/v1.0.1)
